### ISSUE TITLE
@GenerateMocks does not generate correct stub for generic Future return Types

### ISSUE THREAD

**Issue by MaxAuer** [dart-lang/mockito#339]
Hey,

I tried to generate a Mock for the following Function:

```
Future<T> update<T extends Base>(T data);
```

Several classes are extending `Base`.

`flutter pub run build_runner build` does output the following stub:

```
@override
  _i3.Future<T> update<T extends _i4.Base>(T? data) =>
      (super.noSuchMethod(Invocation.method(#update, [data]),
          returnValue: Future.value(null)) as _i3.Future<T>);
```

I then use the following `when` in my test to stub the function

```
when(api.update<Base>(any)).thenThrow(Exception());
```

Which gives the following error `type 'Future<Null>' is not a subtype of type 'Future<Note>' in type cast`. This does make sense as `Future.value(null)` can not be parsed to `Future<T>` which is not nullable. 

The correct behavior should be a value that can be parsed to `Future<T>`.

**Comment by hoc081098** [comment]
Same issue :(

**Comment by hoc081098** [comment]
Trick: edit ` returnValue: Future.value(xxx)` where xxx is value of testing type

**Comment by MaxAuer** [comment]
Maybe worth checking this issue out: [Mocktail-Type-Safety](https://github.com/felangel/mocktail/issues/17).

He could solve this without any code generation. Sure it`s a breaking change but maybe its worth checking out for Mockito.

**Comment by srawlins** [comment]
Mocktail uses strings for mocking, in which case you lose all type safety, typo protection, static analysis, etc. We will not be moving in that direction.

**Comment by MaxAuer** [comment]
Sorry to contradict.

The issue I linked was opened by me because I saw the same issues with Mocktail. It was lacking type safety but works without code generation. Felix found a solution based on lambdas to have type safety without code generation and working with null safety.

He also proposed merging mockito into mocktail: #347

**Comment by kuhnroyal** [comment]
I also have a lot of `Future.value(null)` return values generated that don't compile with strict analyzer settings.
This should be generated as `Future<T>.value(null)` to work with all kinds of `analysis_options.yaml` configurations.

**Comment by srawlins** [comment]
@MaxAuer sorry I got my projects mixed up.

**Comment by srawlins** [comment]
@kuhnroyal can you elaborate? analysis_options.yaml does not affect compilation.

**Comment by kuhnroyal** [comment]
@srawlins You are correct. What I meant to say is, that with `strong-mode` and `language` settings on, the files are red in the IDE as these are not analyzer warnings but errors.

```yaml
analyzer:
  language:
    strict-inference: true
    strict-raw-types: true
  strong-mode:
    implicit-dynamic: false
    implicit-casts: false
``` 

I know that I can exclude the generated files from analysis but it would be nice if the generated code conforms to these setting as it will then also conform to less strict settings. This would make it easier to add to projects.

**Comment by srawlins** [comment]
Thanks @kuhnroyal . That sounds like a separate issue from this issue. I can take a look if you file a separate issue, and include the new static errors you are seeing.

**Comment by srawlins** [comment]
A fix for this issue is available now with Mockito 5.0.9, "fallback generators". Please see the new section in the NULL_SAFETY_README:

https://github.com/dart-lang/mockito/blob/master/NULL_SAFETY_README.md#fallback-generators

**Comment by tomwyr** [comment]
Correct me if I'm wrong here, but since we don't care about actual results of Future objects and fallback generators serve only to make generated code null-safe, couldn't Mockito just return `Completer<T>().future`?

In a similar way methods that return generic streams are currently mocked. The difference here is that instead of a stream that never emits we just want a future that never completes.
